macaroonbakery/httpbakery/client.py

```diff
@@ -169,16 +186,213 @@ def extract_macaroons(headers):
+            if not c.startswith('macaroon-'):
+                continue
+            data = base64.b64decode(cs[c].value)
+            data_as_objs = json.loads(data.decode('utf-8'))
+            ms = [PyMacaroon.deserialize(json.dumps(x),
+                                         serializer=JsonSerializer())
+                  for x in data_as_objs]
+            mss.append(ms)
+    macaroons_header = headers.get('Macaroons')
+    if macaroons_header is not None:
+        data = base64.b64decode(macaroons_header)
         data_as_objs = json.loads(data.decode('utf-8'))
-        ms = [Macaroon.deserialize(json.dumps(x), serializer=JsonSerializer())
+        ms = [PyMacaroon.deserialize(json.dumps(x),
+                                     serializer=JsonSerializer())
               for x in data_as_objs]
         mss.append(ms)
     return mss
+
+
+def acquire_discharge(interaction_methods):
+    def f(ctx, cav, payload):
+        ''' Requesting a discharge macaroon from the caveat location
+        as an HTTP URL.
+        :return Macaroon
+        '''
+        resp = _acquire_discharge_with_token(ctx, cav, payload, None)
+        # TODO Fabrice what is the other http response possible ??
+        if resp.status_code == 200:
+            return macaroonbakery.Macaroon.from_dict(
+                resp.json().get('Macaroon'))
+        cause = Error.deserialize(resp.json())
+        if cause.code != ERR_INTERACTION_REQUIRED:
+            raise DischargeError(resp.json())
+        if cause.info is None:
+            raise DischargeError(
+                'interaction-required response with no info: {}'.format(
+                    resp.json())
+            )
+        loc = cav.location
+        if not loc.endswith('/'):
+            loc = loc + '/'
+        token, m = _interact(ctx, loc, cause, payload, interaction_methods)
+        if m is not None:
+            # We've acquired the macaroon directly via legacy interaction.
+            return m
+        # Try to acquire the discharge again, but this time with
+        # the token acquired by the interaction method.
+        resp = _acquire_discharge_with_token(ctx, cav, payload, token)
+        if resp.status_code == 200:
+            return macaroonbakery.Macaroon.deserialize_json(
+                resp.json().get('Macaroon'))
+        else:
+            raise DischargeError()
+    return f
+
+
+def _acquire_discharge_with_token(ctx, cav, payload, token):
+    req = {}
+    _add_json_binary_field(cav.caveat_id_bytes, req, 'id')
+    if token is not None:
+        _add_json_binary_field(token.value, req, 'token')
+        req['token-kind'] = token.kind
+    if payload is not None:
+        req['caveat64'] = base64.urlsafe_b64encode(payload).rstrip(
+            b'=').decode('utf-8')
+    target = relative_url(cav.location, 'discharge')
+    headers = {
+        BAKERY_PROTOCOL_HEADER: str(LATEST_BAKERY_VERSION)
+    }
+    resp = requests.post(url=target, data=req, headers=headers)
```

Comment:
To match the logic in the Go bakery, this should use the same request logic as the original HTTP call, rather than calling requests.post directly - I guess that means that it should use the same BakeryAuth object when making the request. This makes me think that there should be an object similar to httpbakery.Client that encapsulates the information necessary to make such a call (particularly the client key and the interaction methods), so that _acquire_discharge_with_token and friends can be methods on that object.

Reply:
done